File: TypeScript/ExampleTests/common/PricingTyped.ts
import { PricingString } from "./PricingString.js";
import * as _json from "./json.js";

export class PricingTyped {
  totalPrice: Dollar;

  constructor(totalPrice: Dollar) {
    this.totalPrice = totalPrice;
  }

  static fromStringObj(s: PricingString): PricingTyped {
    return new PricingTyped(
      new Dollar(s.totalPrice)
    );
  }

  toJsonValue(): Record<string, unknown> {
    return {
      totalPrice: this.totalPrice == null ? null : String(this.totalPrice),
    };
  }

  toJSON(): string { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m: unknown): PricingTyped {
    return new PricingTyped(
      new Dollar(_json.asString(_json.requireField(m, "totalPrice"), "totalPrice"))
    );
  }

  static fromJSON(text: string): PricingTyped {
    return PricingTyped.fromJsonValue(_json.parse(text));
  }

  static toJSONList(items: readonly PricingTyped[]): string {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text: string): PricingTyped[] {
    const raw = _json.asArray(_json.parse(text), "PricingTyped") ?? [];
    return raw.map((e) => PricingTyped.fromJsonValue(e));
  }
}

File: Rust/ExampleTests/common/pricing_typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::pricing_string::PricingString;

#[derive(Debug, Clone, Default)]
pub struct PricingTyped {
    pub totalprice: Dollar,
}

impl PricingTyped {
    pub fn from_str_struct(s: &PricingString) -> Self {
        Self {
            totalprice: Dollar::from(s.totalprice.clone()),
        }
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("totalprice".to_string(), json::Value::Str(self.totalprice.to_string())),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            totalprice: Dollar::from(json::as_string(json::require(v, "totalprice")?, "totalprice")?),
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[PricingTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<PricingTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "PricingTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}
